jetty-server/src/main/java/org/eclipse/jetty/server/OptionalSslConnectionFactory.java

```diff
@@ -35,191 +31,76 @@
-                    "Content-Type: text/html\r\n" +
-                    "Content-Length: " + body.length() + "\r\n" +
-                    "Connection: close\r\n" +
-                    "\r\n" +
-                    body;
-            Callback.Completable completable = new Callback.Completable();
-            endPoint.write(completable, ByteBuffer.wrap(response.getBytes(StandardCharsets.US_ASCII)));
-            completable.whenComplete((r, x) -> endPoint.close());
+            // upgrade to the next protocol
+            super.nextProtocol(buffer, endPoint, connector);
         }
         else
         {
-            endPoint.close();
-        }
-    }
-
-    private class OptionalSslConnection extends AbstractConnection implements Connection.UpgradeFrom
-    {
-        private final Connector connector;
-        private final ByteBuffer buffer;
-
-        public OptionalSslConnection(EndPoint endPoint, Connector connector)
-        {
-            super(endPoint, connector.getExecutor());
-            this.connector = connector;
-            this.buffer = BufferUtil.allocateDirect(1536);
-        }
-
-        @Override
-        public void onOpen()
-        {
-            super.onOpen();
-            fillInterested();
-        }
-
-        @Override
-        public void onFillable()
-        {
-            try
-            {
-                while (true)
-                {
-                    int filled = getEndPoint().fill(buffer);
-                    if (filled > 0)
-                    {
-                        // Always have at least 2 bytes.
-                        if (BufferUtil.length(buffer) >= 2)
-                        {
-                            upgrade(buffer);
-                            break;
-                        }
-                    }
-                    else if (filled == 0)
-                    {
-                        fillInterested();
-                        break;
-                    }
-                    else
-                    {
-                        close();
-                        break;
-                    }
-                }
-            }
-            catch (IOException x)
-            {
-                LOG.warn(x);
-                close();
-            }
-        }
-
-        @Override
-        public ByteBuffer onUpgradeFrom()
-        {
-            return buffer;
-        }
-
-        private void upgrade(ByteBuffer buffer)
-        {
-            if (LOG.isDebugEnabled())
-                LOG.debug("Read {}", BufferUtil.toDetailString(buffer));
+            LOG.warn("Detected non-TLS bytes, but no other protocol to upgrade to for {}", endPoint);
 
-            EndPoint endPoint = getEndPoint();
-            if (seemsTLS(buffer))
+            // There are always at least 2 bytes.
+            int byte1 = buffer.get(0) & 0xFF;
+            int byte2 = buffer.get(1) & 0xFF;
+            connector.getByteBufferPool().release(buffer);
+            if (byte1 == 'G' && byte2 == 'E')
```

Comment:
Could this behaviour be moved out of this factory into a `BadHttpRequestConnectionFactory`, so tht the normal detecting logic would choose between an SSL factory and a factory that responds to GE with a 400?

Reply:
Can you detail this? You would like this to be a generic behavior, rather than just something that `OptionalSslConnectionFactory` does?

Reply:
I originally implemented such connection factory but @sbordet did not like it :-)

I'd be happy to re-introduce it.